File: revenue/hive/commercial-laundry-operations/operator_console.py
"""Local browser adapter for the existing laundry operator commands. No public hosting."""
from __future__ import annotations

import argparse
from contextlib import closing
import hmac
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
import json
from pathlib import Path
import secrets
import sqlite3
import sys
import tempfile
from urllib.parse import parse_qs, urlsplit

import operate
from laundry_desk import LaundryDesk, LaundryDeskError, ValidationError

ACTIONS = ("customer", "site", "agreement", "plan", "manifest", "pickup",
           "process", "deliver", "resolve", "invoice-draft")
ASSETS = {"/": ("operator_console.html", "text/html; charset=utf-8"),
          "/operator_console.js": ("operator_console.js", "text/javascript; charset=utf-8")}
FORMATS = {"json": ("application/json", "json"), "csv": ("text/csv", "csv"),
           "markdown": ("text/markdown", "md")}


def overview(database: Path) -> dict:
    desk = LaundryDesk.open_read_only(database)
    with closing(desk._connect()) as conn:
        counts = {name: conn.execute(f"SELECT COUNT(*) FROM {name}").fetchone()[0]
                  for name in ("customers", "sites", "routes")}
        counts["open_exceptions"] = conn.execute(
            "SELECT COUNT(*) FROM exceptions WHERE status='OPEN'").fetchone()[0]
        routes = [dict(row) for row in conn.execute(
            "SELECT route_id,service_date,route_code,state FROM routes "
            "ORDER BY service_date DESC,route_id LIMIT 100")]
        customers = [dict(row) for row in conn.execute(
            "SELECT customer_id,name FROM customers ORDER BY customer_id LIMIT 100")]
        sites = [dict(row) for row in conn.execute(
            "SELECT site_id,customer_id,name FROM sites ORDER BY site_id LIMIT 100")]
    return {"counts": counts, "routes": routes, "customers": customers, "sites": sites,
            "list_limit": 100, "authority": operate.authority()}


class ConsoleServer(ThreadingHTTPServer):
    daemon_threads = True

    def __init__(self, database: Path, port: int):
        self.database = database
        self.token = secrets.token_urlsafe(32)
        self.assets = {url: (Path(__file__).with_name(name).read_bytes(), mime)
                       for url, (name, mime) in ASSETS.items()}
        super().__init__(("127.0.0.1", port), ConsoleHandler)
        self.expected_host = f"127.0.0.1:{self.server_port}"
        self.origin = "http://" + self.expected_host


class ConsoleHandler(BaseHTTPRequestHandler):
    server: ConsoleServer
    server_version = "LaundryConsole/1"
    sys_version = ""

    def setup(self):
        super().setup()
        self.connection.settimeout(10)

    def log_message(self, *_args):
        # No request paths, session credentials or customer data in access logs.
        pass

    def send_bytes(self, status: int, data: bytes, mime: str, filename: str | None = None):
        self.send_response(status)
        self.send_header("Content-Type", mime)
        self.send_header("Content-Length", str(len(data)))
        self.send_header("Cache-Control", "no-store")
        self.send_header("X-Content-Type-Options", "nosniff")
        self.send_header("X-Frame-Options", "DENY")
        self.send_header("Referrer-Policy", "no-referrer")
        self.send_header("Content-Security-Policy", "default-src 'self'; script-src 'self'; "
                         "style-src 'self' 'unsafe-inline'; connect-src 'self'; "
                         "img-src 'none'; frame-ancestors 'none'; base-uri 'none'; form-action 'none'")
        if filename:
            self.send_header("Content-Disposition", f'attachment; filename="{filename}"')
        self.end_headers()
        try:
            self.wfile.write(data)
        except (BrokenPipeError, ConnectionResetError):
            pass  # A lost reply does not undo a committed operation. Retry its exact key.

    def send_json(self, status: int, value: dict):
        self.send_bytes(status, json.dumps(value, ensure_ascii=True).encode(),
                        "application/json; charset=utf-8")

    def permitted(self, *, api: bool, write: bool = False) -> bool:
        if self.headers.get_all("Host") != [self.server.expected_host]:
            self.send_json(403, {"message": "Use the exact loopback address printed at startup."})
            return False
        if api:
            supplied = self.headers.get_all("Authorization") or []
            expected = "Bearer " + self.server.token
            if len(supplied) != 1 or not hmac.compare_digest(
                    supplied[0].encode("utf-8"), expected.encode("ascii")):
                self.send_json(401, {"message": "Unlock with the current console session key."})
                return False
        if write and self.headers.get_all("Origin") != [self.server.origin]:
            self.send_json(403, {"message": "A same-origin browser request is required."})
            return False
        return True

    def do_GET(self):
        if not self.permitted(api=False):
            return
        target = urlsplit(self.path)
        if target.path in self.server.assets and not target.query:
            data, mime = self.server.assets[target.path]
            self.send_bytes(200, data, mime)
            return
        if not self.permitted(api=True):
            return
        try:
            query = parse_qs(target.query, strict_parsing=True, max_num_fields=4)
            if any(len(values) != 1 for values in query.values()):
                raise ValidationError("duplicate query parameter")
            if target.path == "/api/config" and not query:
                fields = {}
                for action in ACTIONS:
                    _, required, optional = operate.COMMANDS[action]
                    fields[action] = {"required": {"operation_key": "str", **{
                        key: kind.__name__ for key, kind in required.items()}},
                        "optional": {key: kind.__name__ for key, kind in optional.items()}}
                self.send_json(200, {"commands": fields})
            elif target.path == "/api/overview" and not query:
                self.send_json(200, overview(self.server.database))
            elif target.path == "/api/route" and set(query) == {"id"}:
                desk = LaundryDesk.open_read_only(self.server.database)
                self.send_json(200, desk.route_snapshot(query["id"][0]))
            elif target.path == "/api/export" and set(query) == {"kind", "id", "format"}:
                kind, identifier, fmt = (query[key][0] for key in ("kind", "id", "format"))
                if kind not in ("route", "customer") or fmt not in FORMATS:
                    raise ValidationError("unsupported export")
                desk = LaundryDesk.open_read_only(self.server.database)
                exports = (desk.render_route_exports(identifier) if kind == "route"
                           else desk.render_customer_exports(identifier))
                mime, extension = FORMATS[fmt]
                self.send_bytes(200, exports[fmt].encode("utf-8"), mime + "; charset=utf-8",
                                f"laundry-{kind}.{extension}")
            else:
                self.send_json(404, {"message": "Unknown console route."})
        except (LaundryDeskError, ValueError) as exc:
            self.send_json(400, {"message": str(exc)})
        except (OSError, sqlite3.Error):
            self.send_json(503, {"message": "The local database could not be read. Check the server terminal."})
            print("Console database read failed.", file=sys.stderr)

    def do_POST(self):
        if not self.permitted(api=True, write=True):
            return
        target = urlsplit(self.path)
        action = target.path.removeprefix("/api/operations/")
        if not target.path.startswith("/api/operations/") or target.query or action not in ACTIONS:
            self.send_json(404, {"message": "Unsupported operation."})
            return
        lengths = self.headers.get_all("Content-Length") or []
        if (len(lengths) != 1 or not 1 <= len(lengths[0]) <= 7 or not lengths[0].isascii() or not lengths[0].isdigit()
                or self.headers.get("Transfer-Encoding") is not None):
            self.send_json(411, {"message": "One fixed Content-Length is required."})
            return
        length = int(lengths[0])
        if not 0 < length <= operate.MAX_INPUT_BYTES:
            self.send_json(413, {"message": "One operation must be at most 1 MiB."})
            return
        if self.headers.get_content_type() != "application/json":
            self.send_json(415, {"message": "Send application/json."})
            return
        try:
            raw = self.rfile.read(length)
            if len(raw) != length:
                raise ValidationError("incomplete request body")
            # Reuse the canonical CLI parser, type rules and engine dispatch unchanged.
            # A temporary input is private to this process and is never a customer file.
            with tempfile.TemporaryDirectory(prefix="laundry-console-") as temporary:
                payload = Path(temporary) / "operation.json"
                payload.write_bytes(raw)
                result = operate.execute(str(self.server.database), action, str(payload))
            self.send_json(200, result)
        except LaundryDeskError as exc:
            self.send_json(409 if type(exc).__name__ in ("StateConflict", "IdempotencyConflict", "InvoiceBlocked")
                           else 400, {"status": "REJECTED", "error_type": type(exc).__name__,
                                      "message": str(exc)})
        except (OSError, sqlite3.Error, OverflowError) as exc:
            # An I/O failure after commit can leave outcome uncertain; never suggest a new key.
            self.send_json(503, {"status": "UNKNOWN", "message":
                "Local I/O failed. Keep the exact operation key and payload; inspect state before retrying."})
            print(f"Console operation I/O failure: {type(exc).__name__}", file=sys.stderr)


def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description="Local laundry operator console; invoices remain DRAFTS")
    parser.add_argument("database", type=Path, help="one operator-owned SQLite database")
    parser.add_argument("--init", action="store_true", help="explicitly create a NEW database; never overwrite")
    parser.add_argument("--port", type=int, default=0, help="loopback port; 0 chooses an available port")
    args = parser.parse_args(argv)
    if not 0 <= args.port <= 65535:
        parser.error("port must be 0..65535")
    try:
        if args.init:
            operate.execute(str(args.database), "init", None)
        operate._regular_database(args.database)
        LaundryDesk.open_read_only(args.database)
        with ConsoleServer(args.database.absolute(), args.port) as server:
            print(f"Open {server.origin}/#key={server.token}", flush=True)
            print("Local session key: keep this address private. Ctrl+C stops the console.", flush=True)
            try:
                server.serve_forever()
            except KeyboardInterrupt:
                pass
    except (LaundryDeskError, OSError, sqlite3.Error) as exc:
        print(f"Console startup failed: {exc}", file=sys.stderr)
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: revenue/hive/commercial-laundry-operations/operator_console.js
"use strict";
const $ = id => document.getElementById(id);
let key = "", epoch = 0, commands = {}, route = null, busy = false, routeRequest = 0;
const keyName = "laundry-console-session-key";
const labels = {operation_key:"Operation key", customer_id:"Customer ID", site_id:"Site ID",
  agreement_id:"Price agreement ID", item_code:"Item code", unit_price_cents:"Unit price in cents",
  active_from:"Active from", active_to:"Active through (optional)", plan_id:"Plan ID",
  route_code:"Route code", weekday:"Weekday (Monday 0 to Sunday 6)", stop_sequence:"Stop sequence",
  service_date:"Service date", stop_id:"Stop ID", linen_counts:"Picked-up counts",
  processed_counts:"Processed-good counts", damaged_counts:"Damaged counts (optional)",
  delivered_counts:"Delivered counts", container_ids:"Container IDs", exception_id:"Exception ID",
  resolution_code:"Resolution code", note:"Supported disposition note", name:"Name"};
function savedKey(value) {
  try {
    if (value === undefined) return sessionStorage.getItem(keyName);
    if (value === null) sessionStorage.removeItem(keyName); else sessionStorage.setItem(keyName, value);
  } catch (_) { /* Private browsing may disable storage; the in-memory session still works. */ }
  return null;
}
const newKey = () => "console." + crypto.randomUUID();
function element(tag, text, className) {
  const node = document.createElement(tag);
  if (text !== undefined) node.textContent = String(text);
  if (className) node.className = className;
  return node;
}
function status(id, message, error=false) {
  $(id).textContent = message;
  $(id).className = "status " + (error ? "error" : "success");
}
async function request(path, options={}, download=false) {
  const generation = epoch;
  const response = await fetch(path, {...options, credentials:"omit", cache:"no-store",
    headers:{Authorization:"Bearer " + key, ...(options.headers || {})}});
  const data = response.ok && download ? await response.blob() : await response.json();
  if (generation !== epoch) throw new Error("The console session changed. Reopen the current workspace.");
  if (!response.ok) throw new Error(data.message || "Local request failed.");
  return data;
}
function datalist(id, rows, field, description) {
  $(id).replaceChildren(...rows.map(row => {
    const option = element("option"); option.value = row[field]; option.label = description(row); return option;
  }));
}
function renderFields() {
  const spec = commands[$("command").value];
  const fields = $("fields"); fields.replaceChildren();
  for (const [name, type] of Object.entries({...spec.required, ...spec.optional})) {
    const label = element("label", labels[name] || name); label.htmlFor = "field-" + name;
    const input = element(type === "dict" || type === "list" ? "textarea" : "input");
    input.id = "field-" + name; input.name = name; input.required = Object.hasOwn(spec.required, name);
    input.dataset.kind = type;
    if (type === "int") { input.type = "number"; input.step = "1"; input.min = "0";
      if (name === "weekday") input.max = "6";
    } else if (name.includes("active_") || name === "service_date") input.type = "date";
    if (type === "dict") input.placeholder = "towel=3\nsheet=2";
    if (type === "list") input.placeholder = "bag-001\nbag-002";
    if (name === "operation_key") input.value = newKey();
    const lists = {customer_id:"customer-options", site_id:"site-options", stop_id:"stop-options"};
    if (lists[name]) input.setAttribute("list", lists[name]);
    label.append(input);
    if (type === "dict" || type === "list") label.append(element("small", type === "dict"
      ? "One item=count per line. Enter observed counts, including zero where appropriate."
      : "One observed container ID per line."));
    fields.append(label);
  }
  status("operation-status", "");
}
function payloadFromForm() {
  const payload = {};
  for (const input of $("fields").querySelectorAll("input, textarea")) {
    const raw = input.value;
    if (!input.required && raw === "") continue;
    const type = input.dataset.kind;
    if (type === "int") {
      if (!/^[0-9]+$/.test(raw) || !Number.isSafeInteger(Number(raw))) throw new Error("Enter a whole-number " + labels[input.name] + ".");
      payload[input.name] = Number(raw);
    } else if (type === "dict") {
      const counts = Object.create(null);
      for (const line of raw.split(/\r?\n/).filter(line => line.trim())) {
        const parts = line.split("=").map(part => part.trim());
        if (parts.length !== 2 || !parts[0] || !/^[0-9]+$/.test(parts[1])
            || !Number.isSafeInteger(Number(parts[1])) || Object.hasOwn(counts, parts[0]))
          throw new Error("Counts need unique item=whole-number lines; nothing was sent.");
        counts[parts[0]] = Number(parts[1]);
      }
      payload[input.name] = counts;
    } else if (type === "list") payload[input.name] = raw.split(/\r?\n/).map(value => value.trim()).filter(Boolean);
    else payload[input.name] = raw;
  }
  return payload;
}
function setBusy(value) {
  busy = value;
  for (const control of $("operation-form").elements) control.disabled = value;
  $("lock").disabled = value;
  for (const button of $("stops").querySelectorAll("button")) button.disabled = value;
}
function prepare(action, field, value) {
  if (busy) return;
  $("command").value = action; renderFields(); $("field-" + field).value = value;
  $("operation-form").scrollIntoView({behavior:"smooth", block:"start"});
  $("field-" + field).focus();
}
async function refresh() {
  const data = await request("/api/overview");
  $("customer-count").textContent = data.counts.customers;
  $("route-count").textContent = data.counts.routes;
  $("exception-count").textContent = data.counts.open_exceptions;
  datalist("route-options", data.routes, "route_id", r => r.service_date + " · " + r.state);
  datalist("customer-options", data.customers, "customer_id", r => r.name);
  datalist("site-options", data.sites, "site_id", r => r.name);
  $("list-note").textContent = `Pickers show at most ${data.list_limit} records each. Older exact IDs can be entered directly.`;
}
function clearRoute() {
  route = null; $("stops").replaceChildren(); $("stop-options").replaceChildren();
  $("route-summary").textContent = "No current route loaded. Open an exact route ID.";
}
async function loadRoute() {
  const sequence = ++routeRequest;
  clearRoute();
  const id = $("route-id").value;
  if (!id) throw new Error("Select or enter an exact route ID.");
  const loaded = await request("/api/route?id=" + encodeURIComponent(id));
  if (sequence !== routeRequest || id !== $("route-id").value) return;
  route = loaded;
  $("route-summary").textContent = `${route.service_date} · ${route.route_code} · ${route.state} · ${route.stops.length} stops`;
  datalist("stop-options", route.stops, "stop_id", s => s.site_name);
  $("stops").replaceChildren();
  for (const stop of route.stops) {
    const card = element("article", undefined, "stop");
    const heading = element("div", undefined, "stop-head");
    heading.append(element("h3", `${stop.sequence} · ${stop.customer_name} / ${stop.site_name}`), element("span", stop.state, "chip"));
    card.append(heading, element("p", stop.stop_id, "mono"));
    const scroll = element("div", undefined, "table-scroll"), table = element("table"), head = element("tr");
    for (const name of ["Item", "Pickup", "Good", "Damage", "Delivered"]) head.append(element("th", name));
    const thead = element("thead"); thead.append(head); table.append(thead);
    const body = element("tbody"), phases = ["pickup_counts", "processed_counts", "damaged_counts", "delivered_counts"];
    const items = [...new Set(phases.flatMap(phase => Object.keys(stop[phase])))].sort();
    for (const item of items) {
      const row = element("tr"); row.append(element("th", item));
      for (const phase of phases) row.append(element("td", Object.hasOwn(stop[phase], item) ? stop[phase][item] : "—"));
      body.append(row);
    }
    table.append(body); scroll.append(table); card.append(scroll);
    card.append(element("p", "Pickup containers: " + (stop.pickup_containers.join(", ") || "Not recorded"), "subtle"),
      element("p", "Delivered containers: " + (stop.delivery_containers.join(", ") || "Not recorded"), "subtle"));
    for (const exception of stop.exceptions) {
      const box = element("div", undefined, "exception");
      box.append(element("strong", `${exception.kind} · ${exception.status}`), element("p", `${exception.item_code}: expected ${exception.expected_qty}, observed ${exception.actual_qty}`), element("p", exception.exception_id, "mono"));
      if (exception.status === "OPEN") {
        const button = element("button", "Review and resolve", "secondary"); button.disabled = busy;
        button.addEventListener("click", () => prepare("resolve", "exception_id", exception.exception_id)); box.append(button);
      } else box.append(element("p", `${exception.resolution_code}: ${exception.resolution_note}`));
      card.append(box);
    }
    if (stop.invoice) card.append(element("div", `DRAFT · USD ${(stop.invoice.total_cents / 100).toFixed(2)} · not issued or paid`, "draft"));
    const next = {MANIFESTED:"pickup", PICKED_UP:"process", PROCESSED:"deliver", DELIVERED:"invoice-draft"}[stop.state];
    if (next) {
      const button = element("button", "Prepare " + next.replace("-", " "), "secondary"); button.disabled = busy;
      button.addEventListener("click", () => prepare(next, "stop_id", stop.stop_id)); card.append(button);
    }
    $("stops").append(card);
  }
  status("route-status", "Current route loaded. Review all open exceptions before drafting.");
}
async function download(kind) {
  const id = $(kind === "route" ? "route-id" : "export-customer-id").value;
  if (!id) throw new Error("Enter the exact " + kind + " ID first.");
  const format = $(kind + "-format").value;
  const query = new URLSearchParams({kind, id, format});
  const blob = await request("/api/export?" + query, {}, true);
  const link = element("a"); link.href = URL.createObjectURL(blob);
  link.download = `laundry-${kind}.${format === "markdown" ? "md" : format}`;
  document.body.append(link); link.click(); link.remove();
  setTimeout(() => URL.revokeObjectURL(link.href), 1000);
  status("route-status", "Downloaded native " + kind + " " + format + ". Nothing was sent to a customer.");
}
function lock() {
  epoch++; routeRequest++; key = ""; savedKey(null); clearRoute();
  $("workspace").hidden = true; $("lock").hidden = true; $("unlock-panel").hidden = false;
  $("fields").replaceChildren(); $("stops").replaceChildren(); $("receipt").textContent = "";
  for (const id of ["route-options", "customer-options", "site-options"]) $(id).replaceChildren();
  for (const id of ["customer-count", "route-count", "exception-count"]) $(id).textContent = "0";
  $("route-id").value = ""; $("export-customer-id").value = "";
  status("route-status", ""); status("operation-status", "");
  $("session-key").value = ""; $("session-key").focus();
}
async function unlock(value) {
  epoch++; key = value;
  try {
    commands = (await request("/api/config")).commands;
    await refresh(); renderFields(); savedKey(key);
    $("unlock-panel").hidden = true; $("workspace").hidden = false; $("lock").hidden = false;
    $("session-key").value = "";
  } catch (error) { lock(); status("unlock-status", error.message, true); }
}
$("unlock-form").addEventListener("submit", event => { event.preventDefault(); unlock($("session-key").value.trim()); });
$("lock").addEventListener("click", lock);
$("command").addEventListener("change", renderFields);
$("new-key").addEventListener("click", () => { $("field-operation_key").value = newKey(); status("operation-status", "New key selected. This is a new operation, not a retry."); });
$("operation-form").addEventListener("submit", async event => {
  event.preventDefault(); if (busy) return;
  try {
    const action = $("command").value, payload = payloadFromForm(); setBusy(true);
    status("operation-status", "Recording locally. Keep this key until the result is known.");
    const result = await request("/api/operations/" + action, {method:"POST", headers:{"Content-Type":"application/json"}, body:JSON.stringify(payload)});
    $("receipt").textContent = JSON.stringify(result, null, 2);
    status("operation-status", `${result.status} · ${result.operation_key}. This key remains available for an exact replay.`);
    if (action === "manifest") $("route-id").value = result.result.route_id;
    try { await refresh(); if ($("route-id").value) await loadRoute(); }
    catch (error) { status("route-status", "Operation acknowledged; refreshing the view failed: " + error.message, true); }
  } catch (error) { status("operation-status", error.message + " Keep the same key and fields for a retry.", true); }
  finally { setBusy(false); }
});
for (const [id, action] of [["open-route", loadRoute], ["refresh", async () => { await refresh(); if ($("route-id").value) await loadRoute(); }], ["export-route", () => download("route")], ["export-customer", () => download("customer")]]) {
  $(id).addEventListener("click", () => action().catch(error => { status("route-status", error.message, true); }));
}
$("route-id").addEventListener("input", () => { routeRequest++; clearRoute(); });
const incoming = new URLSearchParams(location.hash.slice(1)).get("key");
if (location.hash) history.replaceState(null, "", location.pathname);
const initial = incoming || savedKey();
if (initial) unlock(initial);
